tooling/src/hypothesistooling/projects/hypothesispython.py

```diff
@@ -0,0 +1,242 @@
+
+assert os.path.exists(PYTHON_SRC)
+
+
+__version__ = None
+__version_info__ = None
+
+VERSION_FILE = os.path.join(PYTHON_SRC, 'hypothesis/version.py')
+
+with open(VERSION_FILE) as o:
+    exec(o.read())
+
+
+def has_release():
+    return os.path.exists(RELEASE_FILE)
+
+
+def has_source_changes():
+    return tools.has_changes([PYTHON_SRC])
+
+
+CHANGELOG_ANCHOR = re.compile(r"^\.\. _v\d+\.\d+\.\d+:$")
+CHANGELOG_BORDER = re.compile(r"^-+$")
+CHANGELOG_HEADER = re.compile(r"^\d+\.\d+\.\d+ - \d\d\d\d-\d\d-\d\d$")
+RELEASE_TYPE = re.compile(r"^RELEASE_TYPE: +(major|minor|patch)")
+
+
+MAJOR = 'major'
+MINOR = 'minor'
+PATCH = 'patch'
+
+VALID_RELEASE_TYPES = (MAJOR, MINOR, PATCH)
+
+
+def parse_release_file():
+    with open(RELEASE_FILE) as i:
+        release_contents = i.read()
+
+    release_lines = release_contents.split('\n')
+
+    m = RELEASE_TYPE.match(release_lines[0])
+    if m is not None:
+        release_type = m.group(1)
+        if release_type not in VALID_RELEASE_TYPES:
+            print('Unrecognised release type %r' % (release_type,))
+            sys.exit(1)
+        del release_lines[0]
+        release_contents = '\n'.join(release_lines).strip()
+    else:
+        print(
+            'RELEASE.rst does not start by specifying release type. The first '
+            'line of the file should be RELEASE_TYPE: followed by one of '
+            'major, minor, or patch, to specify the type of release that '
+            'this is (i.e. which version number to increment). Instead the '
+            'first line was %r' % (release_lines[0],)
+        )
+        sys.exit(1)
+
+    return release_type, release_contents
+
+
+def update_changelog_and_version():
+    global __version_info__
+    global __version__
+
+    with open(CHANGELOG_FILE) as i:
+        contents = i.read()
+    assert '\r' not in contents
+    lines = contents.split('\n')
+    assert contents == '\n'.join(lines)
+    for i, l in enumerate(lines):
+        if CHANGELOG_ANCHOR.match(l):
+            assert CHANGELOG_BORDER.match(lines[i + 2]), repr(lines[i + 2])
+            assert CHANGELOG_HEADER.match(lines[i + 3]), repr(lines[i + 3])
+            assert CHANGELOG_BORDER.match(lines[i + 4]), repr(lines[i + 4])
+            beginning = '\n'.join(lines[:i])
+            rest = '\n'.join(lines[i:])
+            assert '\n'.join((beginning, rest)) == contents
+            break
+
+    release_type, release_contents = parse_release_file()
+
+    new_version = list(__version_info__)
+    bump = VALID_RELEASE_TYPES.index(release_type)
+    new_version[bump] += 1
+    for i in range(bump + 1, len(new_version)):
+        new_version[i] = 0
+    new_version = tuple(new_version)
+    new_version_string = '.'.join(map(str, new_version))
+
+    __version_info__ = new_version
+    __version__ = new_version_string
+
+    with open(VERSION_FILE) as i:
+        version_lines = i.read().split('\n')
+
+    for i, l in enumerate(version_lines):
+        if 'version_info' in l:
```

Comment:
minor: would `if l.startswith('__version_info__'):` be a tighter check here?

(I think this is copied from previous code, I don’t actually remember)

Reply:
It is indeed copied from previous code. I'd rather keep from adding too many changes to the underlying code in this PR.